### DISCUSSION TITLE
Please help with multicast setup over switch as router and firewall

### DISCUSSION THREAD

**Discussion by eriksl** [troglobit/pimd#238] [Q&A]
Hello all,

I trying to send simple multicast traffic from my PC to my (ESP8266) microcontrollers. In the past this has worked so I am fairly confident neither the sending client nor the microcontrollers are the problem.

Now some time ago I changed my network to be fully layer 3. This has been working quite like expected, no problems there. But the multicast, I can't get it working. Formerly I had "smclient" running on the firewall which would forward all respective multicast traffic from the "intern" vlan to the "esp" vlan, works.

Now I can't do this anymore, because the switch interfaces need to be instructed to forward the multicast traffic to the firewall and the receiving client on the switch.

Please see the current situation in the drawing below:

![netwerk](https://user-images.githubusercontent.com/1038982/219726933-b40af89d-f1f8-499c-bb62-337ef8b01ed7.png)

For the time being I have enabled pim on both the firewall (pimd) and the switch VRF's (2x). I don't particularly want to use pim, but I cannot seem to add static multicast routes on the switch... I configured pimd in a very simple way on the firewall, and by golly, all instances seem to see and recognise each other. But it doesn't work.

A software client on computer eos is sending traffic to 239.19.255.1. For some reason it understands that it should be send towards the default gateway and it sends up at the link to ge-0/0/26 of the switch. The switch simply drops the traffic. It also doesn't mention a multicast route in that vrf. It doesn't even look like pimd is sending to the VRF 16 instance it likes to receive this traffic.

From the other side, the receiving clients, the igmp table at the switch shows the multicast group 239.19.255.1 does exist and is requested by one or more clients in vlan 219. Pimd also mentions this fact in the debug output, but no multicast route is created (ip mroute) and no request is made to the VRF 16 switch to deliver the traffic.

Can someone please help me get this on the rails? Unfortunately there is an overwhelming lack of documentation on this subject internet, so I am almost completely in the dark, even though I think I have a basic grasp of the subject.

Please note that although a firewall is present in the path, none of the respective traffic is dropped or rejected (checked the logging very frequently).

Note: I now see there is also a pim-dense-daemon, would that be a better option (because no pruning would be required, the "route" is static and always known in advance, I just need to get my switch to do the forwarding, based on either pim or igmp instructions).

Two of my most prominent questions:
- could this be done all without pim, using IGMP and "something" proxying them on the firewall (smcroute, igmproxy)? The route of the packet is completely static and need only to be duplicated at the very last hop.
- how many rendez-vous points are required for this setup? Would one RP suffice, or would I need one for every VRF? How would both RPs find each other? Could this be done using one pimd instance or would it require two pimd instances?

Thank you all very much for any, just **any** input!

**Comment by cmongolang** [comment]
Hi Erik, I also work with healthcare multicast.
You need each router interface along the path to run the PIM service. If this is possible, you would need a single RP for PIM-SM. The IP address RP role is then entered manually in each router, or shared bya BSR instance. 
For PIM dense mode, no RP or BSR is used (flood then prune).
FYI I've found RouterOS V6 very easy to get up and running with PIM-SM. It's the OS for Mikrotik devices but also runs on x86, and is very handy to run in GNS3. RouterOS V7 multicast still seems flaky.

**Reply by eriksl** [reply]
Hi cmongolong,

The situation that I am trying now is like this (with various variations):
- the firewall runs pimd
- the switch runs two instances of pim, each VRF has one.

I configured the switch to have a static RP (which resides in the firewall), I tried both two RP's (each per VRF) or one RP, but it doesn't matter. At the moment I am trying "BSR" mode where the firewall advertises itself and the switch has it's BSR role "disabled". This all works, they can see each other, but the end result is still the same.

FWIW my current configuration of pimd:

```
phyint vlan16 enable altnet 172.16.0.0/16 igmpv2
phyint vlan19 enable altnet 172.19.0.0/16 igmpv2
bsr-candidate priority 255
rp-candidate priority 255 group-prefix 239.19.0.0/16
```

I am not sure if the "altnet" is required. Formerly I also used

```
rp-address 172.16.193.0    239.0.0.0/8
```
instead of "bsr-candidate" and "rp-candidate".

The configuration on the switch (VRF 19):
```
rp {
    local {
        disable;
    }
}
interface vlan.19 {
    mode sparse;
    version 2;
}
```
The configuration on the switch (VRF 16):
```
rp {
    local {
        disable;
    }
}
interface vlan.16 {
    mode sparse;
    version 2;
}
```
I also tried (but no difference) (two RP's):
```
rp static address 172.16.255.2 # VRF16
rp static address 172.19.255.2 # VRF17
```
And I tried (one RP, change the pimd config accordingly):
```
rp static address 172.19.255.2 # VRF16
rp static address 172.19.255.2 # VRF17
```

**Comment by eriksl** [comment]
This is interesting, the debug output says this, but ip mroute show nothing:

```
Multicast Routing Table ======================================================
----------------------------------- (*,G) ------------------------------------
Source           Group            RP Address       Flags
---------------  ---------------  ---------------  ---------------------------
INADDR_ANY       239.19.255.1     172.19.255.2     WC RP
Joined   oifs: ...j..              
Pruned   oifs: ......              
Leaves   oifs: ......              
Asserted oifs: ......              
Outgoing oifs: ...o..              
Incoming     : .....I              

TIMERS:  Entry    JP    RS  Assert VIFS:  0  1  2  3  4  5
           160    15     0       0        0  0  0 160  0  0
----------------------------------- (S,G) ------------------------------------
----------------------------------- (*,G) ------------------------------------
Source           Group            RP Address       Flags
---------------  ---------------  ---------------  ---------------------------
INADDR_ANY       239.19.255.2     172.19.255.2     WC RP
Joined   oifs: ...j..              
Pruned   oifs: ......              
Leaves   oifs: ......              
Asserted oifs: ......              
Outgoing oifs: ...o..              
Incoming     : .....I              
```

**Reply by troglobit** [reply]
When routing multicast you need to ensure the TTL is > 1.  It's actually more complicated than that, but lets start there since it's the most common problem.

And yeah, sure you can use SMCRoute (I maintain that too), or igmpproxy and similar.  Each have their advantages, I personally like pimd-dense since it "just works", basically, and you don't need much logi c except working unicast routing, which people seem to be more comfortable to set up.  With SMCRoute you need to know the IP and group that the multicast sender use to really leverage the static nature of that daemon.   But only you know the requirements that matters to your use-case.

**Comment by eriksl** [comment]
Hi, thank you for your information.

The TTL I actually already took care of in the previous setup, as it still would have been routed by the firewall, I set the TTL to 4 which, I think, should be enough in this situation?

I think the problem i am running into is that my switch does not understand static multicast routes. So one way or another it needs to be instructed to forward the traffic that is coming into ge-0/0/26 (eos) to the firewall. I was hoping pimd would take care of that using the pim protocol, but there are never going pim join/prune requests for this group to the switch, the switch does not know of the group and the packets coming in into ge-0/0/26 are not forwarded.

So how do I get the switch to add entries into it's multicast route table? I guess when not using pim, this could be achieved with IGMP. Would smcroute also send IGMP join packets?

I do not really care if the remaining setup is static, dynamic, configuration per group or all, I'm already happy if I get it working. The fact that there is little information about pim and igmp on JunOS doesn't really help too. E.g. I can enable "IGMP" on an interface but what does it do? Would that also be required when using PIM or would it interfere? The interfaces having "PIM" enabled would that only be the interfaces that have a peer PIM router or also the ones having clients only? So you see, a lot of unknown variables :-(

In the meantime I will have a look at igmp proxy, pimdd and compile pimd myself (I am now using a distribution version, which seems a bit old).

**Reply by troglobit** [reply]
OK, so here's how it's supposed to work.

 1. On the LAN multicast should be flooded to multicast routers
 2. PIM traffic is (usually) only acted upon by other PIM routers
 3. On the receiving LAN side the consumer joins the group using IGMP

Switches that support IGMP snooping must either recognize PIM routers, and mark ports in the direction of the router as a multicast router port, or have a way to set a static multicast router port.  Another way, that is recommended, is to have the router as the IGMP querier.  This is usually achieved by making sure the router has a lower IP address.  Switches can also be set in IGMP proxy mode to prevent them from ever becoming the elected IGMP querier.  (Per RFC the querier is where multicast is forwarded to.)

In mrouted we recently added support for the router to join groups on the sender side to open up forwarding in any intermediate switch (because they are often buggy).  On the sender side there is already a "fake receiver" setting in mrouted.  Please note, the latest feature addition has not been released yet.  There are plans to port these mrouted-specific features to pimd.

**Reply by eriksl** [reply]
Hi, I still have the question, how do I get the first hop (in my case the EX3300 in VRF 16) to forward the/any multicast traffic to the next hop in the chain (in this case the Linux firewall). I don't really have a "lan" as each client has it's own L3 ip range/port on the switch (not even a vlan). This is very Juniper specific, I admit, but with some keywords I might find the specific configuration.

**Reply by troglobit** [reply]
It sounds like you don't have a problem with pimd but rather forwarding to the router running pimd, so maybe you should ask your questions in a Juniper forum?

**Reply by eriksl** [reply]
That really is part of the problem indeed. I already asked on the Juniper community forum and on Quora (both English and Dutch) but apparently nobody knows (or wants to tell me ;-)).

What I find interesting (on the firewall/Linux side, that whatever I do, IGMP joins are never sent to the switch, so I can imagine it doesn't work. I am now experimenting with smcroute, the "mgroup" keyword suggests it also joins the group on the incoming interface (vlan 16 in this case) but I never see any IGMP joins on vlan 16!
```
phyint vlan16 enable mrdisc
phyint vlan19 enable mrdisc
mgroup from vlan16 source 172.16.26.2 group 239.19.255.1
mroute from vlan16 source 172.16.26.2 group 239.19.255.1 to vlan19
```
Now when I set a specific source I do at least see a multicast route appearing:
```
172.16.26.2,239.19.255.1)       Iif: vlan16     Oifs: vlan19  State: resolved
```

**Reply by eriksl** [reply]
Update: with a new version of smcroute I do see the joins. But still no traffic from the switch. Next problem to solve ;-)

**Reply by eriksl** [reply]
I am getting pretty desperate now. Nothing with either smcroute or pimdd too. Simple question, if multicast traffic enters a router (in this case ge-0/0/26) and there is no multicast routing active (or a _dense_ one), it should simply forward/flood the traffic to all other interfaces, right? I am not seeing anything, no traffic at all. Maybe it has to do with that face that the firewall is connected through an L3 vlan interface, who knows. I think it should make no difference, there are only two hosts in it.

**Reply by troglobit** [reply]
Only dense mode multicast routing protocols are flood-and-prune.  PIM-SM/SSM are sparse mode and do not have this behavior.  In PIM-SM incoming multicast is forwarded to the rendez-vous point (RP) for that group.  An IGMP join on the receiver LAN is translated to a PIM Join towards that RP.  

If you want to run something else on your Linux firewall, e.g. SMCRoute, you need to send an IGMP join towards the Juniper on the source LAN. This you seem to have done above with the `mgroup` statement.  If the Juniper switch still does not forward the multicast, there's not much I or this project can help you with.

**Reply by eriksl** [reply]
Yes I see, I already feared this. I have spend quite a few hours now and whatever I do, the multicast traffic never gets forwarded (including pim dense mode).

I am going to have one last go at trying to arrange it all statically and if that doesn't work, I'll have to leave this project until I find someone very knowledgeable on Juniper.

Thanks for you help!

**Reply by cmongolang** [reply]
One other point, if you disable igmp snooping on the switch, it should treat the multicast as broadcast within the single VLAN.

**Reply by eriksl** [reply]
igmp snooping is off, because I don't have any L2 domains on the switch. It's acting purely as a router. I have a few vlan interfaces configured because:
- between the switch and the firewall: if I create subinterfaces on the interfaces towards the firewall, the switch crashes very hard, it's a bug with having subinterfaces in different routing instances ("vrf's"). So I am using point-to-point L3 vlans which work fine and in the end do the same job.
- between the switch and the access points: a small piece of L2 domain is required here because I am using more than one ap and clients need to be able to roam. The first L3 point is at the switch, it's the L3 vlan interface.

So there is never the situation where two clients are in the same vlan, which would call for the use of igmp snooping.

What I want to do is that the switch forwards (as in: _routing_, not as in _bridging_) all non-link-scope (224/24) multicast to another interface (which happens to be a vlan interface, but is purely L3). Apparently Juniper does not facilitate this is in any way. It should be possible with pim (dense/sparse) or even igmp proxying, but nothing is happening, the multicast traffic is simply dropped.

**Comment by eriksl** [comment]
This is the packet I am sending to the switch (ge-0/0/26):
```
Frame 6: 45 bytes on wire (360 bits), 45 bytes captured (360 bits) on interface eth0, id 0
    Interface id: 0 (eth0)
        Interface name: eth0
    Encapsulation type: Ethernet (1)
    Arrival Time: Feb 18, 2023 09:34:53.148336154 CET
    [Time shift for this packet: 0.000000000 seconds]
    Epoch Time: 1676709293.148336154 seconds
    [Time delta from previous captured frame: 1.002120036 seconds]
    [Time delta from previous displayed frame: 1.002120036 seconds]
    [Time since reference or first frame: 5.015645437 seconds]
    Frame Number: 6
    Frame Length: 45 bytes (360 bits)
    Capture Length: 45 bytes (360 bits)
    [Frame is marked: False]
    [Frame is ignored: False]
    [Protocols in frame: eth:ethertype:ip:udp:data]
Ethernet II, Src: 48:4d:7e:da:d0:75, Dst: 01:00:5e:13:ff:02
    Destination: 01:00:5e:13:ff:02
        Address: 01:00:5e:13:ff:02
        .... ..0. .... .... .... .... = LG bit: Globally unique address (factory default)
        .... ...1 .... .... .... .... = IG bit: Group address (multicast/broadcast)
    Source: 48:4d:7e:da:d0:75
        Address: 48:4d:7e:da:d0:75
        .... ..0. .... .... .... .... = LG bit: Globally unique address (factory default)
        .... ...0 .... .... .... .... = IG bit: Individual address (unicast)
    Type: IPv4 (0x0800)
Internet Protocol Version 4, Src: 172.16.26.2, Dst: 239.19.255.2
    0100 .... = Version: 4
    .... 0101 = Header Length: 20 bytes (5)
    Differentiated Services Field: 0x00 (DSCP: CS0, ECN: Not-ECT)
        0000 00.. = Differentiated Services Codepoint: Default (0)
        .... ..00 = Explicit Congestion Notification: Not ECN-Capable Transport (0)
    Total Length: 31
    Identification: 0x96b2 (38578)
    Flags: 0x40, Don't fragment
        0... .... = Reserved bit: Not set
        .1.. .... = Don't fragment: Set
        ..0. .... = More fragments: Not set
    ...0 0000 0000 0000 = Fragment Offset: 0
    Time to Live: 4
    Protocol: UDP (17)
    Header Checksum: 0x2bf3 [validation disabled]
    [Header checksum status: Unverified]
    Source Address: 172.16.26.2
    Destination Address: 239.19.255.2
User Datagram Protocol, Src Port: 47384, Dst Port: 24
    Source Port: 47384
    Destination Port: 24
    Length: 11
    Checksum: 0xb445 [unverified]
    [Checksum Status: Unverified]
    [Stream index: 0]
    [Timestamps]
        [Time since first frame: 5.015645437 seconds]
        [Time since previous frame: 1.002120036 seconds]
    UDP payload (3 bytes)
Data (3 bytes)
```
